src/pudl/workspace/datastore.py

```diff
@@ -240,16 +236,24 @@ def _fetch_from_url(self, url: str) -> requests.Response:
 
     def _doi_to_url(self, doi: str) -> str:
         """Returns url that holds the datapackage for given doi."""
-        match = re.search(r"zenodo.([\d]+)", doi)
-        if match is None:
-            raise ValueError(f"Invalid doi {doi}")
+        match = re.search(r"(10\.5072|10\.5281)/zenodo.([\d]+)", doi)
 
-        zen_id = int(match.groups()[0])
-        return f"{self._api_root}/deposit/depositions/{zen_id}"
+        if match is None:
+            raise ValueError(f"Invalid Zenodo DOI: {doi}")
+
+        doi_prefix = match.groups()[0]
+        zenodo_id = match.groups()[1]
+        if doi_prefix == "10.5072":
+            api_root = self.API_ROOT["sandbox"]
+        elif doi_prefix == "10.5281":
+            api_root = self.API_ROOT["production"]
+        else:
+            raise ValueError(f"Invalid Zenodo DOI: {doi}")
+        return f"{api_root}/deposit/depositions/{zenodo_id}"
```

Comment:
I'm sure there's a more eloquent way of switching between production and sandbox on a per-dataset basis (rather than the whole instance of the class being tied to one or the other) but this seems relatively self-contained and not terrible for the moment.

Reply:
I messed around with creating a DOI class:

```py
class ZenodoDoi(BaseModel):
    """A class defining useful validations and methods for working with Zenodo DOIs."""

    doi: constr(regex=r"^10\.(5072|5281)/zenodo\.[\d]+$")  # noqa: F722

    def __str__(self: Self) -> str:
        """String representation of the DOI"""
        return self.doi

    @property
    def is_prod(self: Self) -> bool:
        """Return True if DOI is from Zenodo production server, False otherwise."""
        if self.doi.startswith("10.5281/zenodo"):
            return True
        else:
            assert self.doi.startswith("10.5072/zenodo")
            return False

    @property
    def token(self: Self) -> str:
        """Zenodo read-only personal access token corresponding to this DOI.

        Zenodo tokens recorded here should have read-only access to our archives.
        Including them here is correct in order to allow public use of this tool, so
        long as we stick to read-only keys.
        """
        # Read-only personal access tokens for pudl@catalyst.coop:
        if self.is_prod:
            return "KXcG5s9TqeuPh1Ukt5QYbzhCElp9LxuqAuiwdqHP0WS4qGIQiydHn6FBtdJ5"
        else:
            return "qyPC29wGPaflUUVAv1oGw99ytwBqwEEdwi4NuUrpwc3xUcEwbmuB4emwysco"

    @property
    def zenodo_id(self: Self) -> str:
        """The Zenodo deposition ID, extracted from the DOI."""
        match = re.search(r"(10\.5072|10\.5281)/zenodo.([\d]+)", self.doi)
        return match.groups()[1]

    @property
    def api_root(self: Self) -> HttpUrl:
        """Return appropriate production or sandbox Zenodo API root URL."""
        if self.is_prod:
            return "https://zenodo.org/api"
        else:
            return "https://sandbox.zenodo.org/api"

    @property
    def url(self: Self) -> HttpUrl:
        """Zenodo URL corresponding to this DOI."""
        return f"{self.api_root}/deposit/depositions/{self.zenodo_id}"
```